File: backends/ruby-rails/app/services/stackverse/input_validation.rb
module Stackverse
  module InputValidation
    require "uri"

    UUID_PATTERN = /\A[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\z/i
    TAG_PATTERN = /\A[a-z0-9-]{1,30}\z/
    MESSAGE_KEY_PATTERN = /\A[a-z0-9-]+(\.[a-z0-9-]+)*\z/
    LANGUAGE_PATTERN = /\A[a-z]{2}\z/
    VISIBILITIES = %w[private public].freeze
    REPORT_REASONS = %w[spam offensive broken-link other].freeze
    REPORT_STATUSES = %w[open dismissed actioned].freeze

    module_function

    def parse_uuid(value)
      Errors.not_found unless value.to_s.match?(UUID_PATTERN)
      value.downcase
    end

    def max_length(value, limit, name)
      Errors.bad_request("#{name} must be at most #{limit} characters") if value && value.length > limit
    end

    def validate_bookmark(body)
      input = body.is_a?(Hash) ? body : {}
      validator = Validator.new
      url = input["url"].is_a?(String) ? input["url"].strip : ""
      if url.empty?
        validator.reject("url", "validation.url.required")
      else
        validator.check(url.length <= 2000 && http_url?(url), "url", "validation.url.invalid")
      end

      title = input["title"].is_a?(String) ? input["title"].strip : ""
      validator.check(!title.empty?, "title", "validation.title.required")
      validator.check(title.length <= 200, "title", "validation.title.too-long")

      notes = input["notes"].is_a?(String) ? input["notes"] : nil
      validator.check(notes.to_s.length <= 4000, "notes", "validation.notes.too-long")

      raw_tags = input["tags"].is_a?(Array) ? input["tags"] : []
      tags = raw_tags.map { |tag| tag.to_s.strip.downcase }.uniq
      validator.check(tags.length <= 10, "tags", "validation.tags.too-many")
      validator.check(tags.all? { |tag| tag.match?(TAG_PATTERN) }, "tags", "validation.tag.invalid")

      visibility = input.fetch("visibility", "private")
      Errors.bad_request("unknown visibility: #{visibility}") unless VISIBILITIES.include?(visibility)

      validator.throw_if_invalid
      { url: url, title: title, notes: notes, tags: tags, visibility: visibility }
    end

    def validate_query_tags(tags)
      normalized = tags.map { |tag| tag.strip.downcase }
      validator = Validator.new
      validator.check(normalized.all? { |tag| tag.match?(TAG_PATTERN) }, "tag", "validation.tag.invalid")
      validator.throw_if_invalid
      normalized
    end

    def validate_report(body)
      input = body.is_a?(Hash) ? body : {}
      validator = Validator.new
      reason = input["reason"]
      validator.check(reason.is_a?(String) && REPORT_REASONS.include?(reason), "reason", "validation.report.reason.invalid")
      comment = input["comment"].is_a?(String) ? input["comment"] : nil
      validator.check(comment.to_s.length <= 1000, "comment", "validation.report.comment.too-long")
      validator.throw_if_invalid
      { reason: reason, comment: comment }
    end

    def validate_report_status(value)
      return nil if value.nil?

      Errors.bad_request("unknown status: #{value}") unless REPORT_STATUSES.include?(value)
      value
    end

    def validate_resolution(body)
      input = body.is_a?(Hash) ? body : {}
      validator = Validator.new
      resolution = input["resolution"]
      validator.check(resolution.is_a?(String) && REPORT_STATUSES.include?(resolution), "resolution", "validation.resolution.invalid")
      note = input["note"].is_a?(String) ? input["note"] : nil
      validator.check(note.to_s.length <= 1000, "note", "validation.resolution.note.too-long")
      validator.throw_if_invalid
      [resolution, note]
    end

    def validate_bookmark_status(body)
      input = body.is_a?(Hash) ? body : {}
      validator = Validator.new
      status = input["status"]
      validator.check(%w[active hidden].include?(status), "status", "validation.bookmark-status.invalid")
      note = input["note"].is_a?(String) ? input["note"] : nil
      validator.check(note.to_s.length <= 1000, "note", "validation.bookmark-status.note.too-long")
      validator.throw_if_invalid
      [status, note]
    end

    def validate_message(body)
      input = body.is_a?(Hash) ? body : {}
      validator = Validator.new
      key = input["key"].is_a?(String) ? input["key"].strip : ""
      validator.check(key.match?(MESSAGE_KEY_PATTERN) && key.length <= 150, "key", "validation.message.key.invalid")
      language = input["language"].is_a?(String) ? input["language"].strip : ""
      validator.check(language.match?(LANGUAGE_PATTERN), "language", "validation.message.language.invalid")
      text = input["text"].is_a?(String) ? input["text"] : ""
      validator.check(!text.empty?, "text", "validation.message.text.required")
      validator.check(text.length <= 2000, "text", "validation.message.text.too-long")
      description = input["description"].is_a?(String) ? input["description"] : nil
      validator.check(description.to_s.length <= 1000, "description", "validation.message.description.too-long")
      validator.throw_if_invalid
      { key: key, language: language, text: text, description: description }
    end

    def validate_user_status(body, target_username, actor_username)
      input = body.is_a?(Hash) ? body : {}
      status = input["status"]
      Errors.bad_request("status is required") unless %w[active blocked].include?(status)
      reason = input["reason"].is_a?(String) ? input["reason"].strip : nil
      if status == "blocked"
        validator = Validator.new
        validator.check(reason.present?, "reason", "validation.block.reason.required")
        validator.check(reason.to_s.length <= 1000, "reason", "validation.block.reason.too-long")
        validator.throw_if_invalid
        Errors.conflict("Admins cannot block themselves.") if target_username == actor_username
      end
      [status, reason]
    end

    def parse_datetime(value, name)
      return nil if value.nil?

      Time.iso8601(value)
    rescue ArgumentError
      Errors.bad_request("#{name} must be an RFC 3339 date-time")
    end

    def http_url?(value)
      uri = URI.parse(value)
      %w[http https].include?(uri.scheme) && uri.host.present?
    rescue URI::InvalidURIError
      false
    end
  end
end

File: backends/scala-http4s/src/main/scala/dev/stackverse/http4s/InputValidation.scala
package dev.stackverse.http4s

import io.circe.JsonObject

import java.net.URI
import scala.util.Try

object InputValidation {
  private val TagPattern = "^[a-z0-9-]{1,30}$".r
  private val KeyPattern = "^[a-z0-9-]+(\\.[a-z0-9-]+)*$".r
  private val LanguagePattern = "^[a-z]{2}$".r

  case class BookmarkInput(url: String, title: String, notes: Option[String], tags: Seq[String], visibility: String)
  case class MessageInput(key: String, language: String, text: String, description: Option[String])
  case class ReportInput(reason: String, comment: Option[String])
  case class ListFilters(tags: Seq[String], q: Option[String], visibility: Option[String])

  def validateBookmarkInput(body: JsonObject): BookmarkInput = {
    val validator = Validator()
    val url = body("url").flatMap(_.asString).map(_.trim).getOrElse("")
    if (url.isEmpty) validator.reject("url", "validation.url.required")
    else validator.check(url.length <= 2000 && isHttpUrl(url), "url", "validation.url.invalid")
    val title = body("title").flatMap(_.asString).map(_.trim).getOrElse("")
    validator.check(title.nonEmpty, "title", "validation.title.required")
    validator.check(title.length <= 200, "title", "validation.title.too-long")
    val notes = body("notes").flatMap(_.asString)
    validator.check(notes.forall(_.length <= 4000), "notes", "validation.notes.too-long")
    val rawTags = body("tags").flatMap(_.asArray).getOrElse(Vector.empty)
    val tags = rawTags.map(value => value.asString.getOrElse(value.noSpaces)).map(_.trim.toLowerCase).distinct
    validator.check(tags.length <= 10, "tags", "validation.tags.too-many")
    validator.check(tags.forall(tag => TagPattern.matches(tag)), "tags", "validation.tag.invalid")
    val visibility = body("visibility").flatMap(_.asString).getOrElse("private")
    if (visibility != "private" && visibility != "public") throw BadRequestProblem(s"unknown visibility: $visibility")
    validator.throwIfInvalid()
    BookmarkInput(url, title, notes, tags, visibility)
  }

  def validateMessageInput(body: JsonObject): MessageInput = {
    val validator = Validator()
    val key = body("key").flatMap(_.asString).map(_.trim).getOrElse("")
    validator.check(KeyPattern.matches(key) && key.length <= 150, "key", "validation.message.key.invalid")
    val language = body("language").flatMap(_.asString).map(_.trim).getOrElse("")
    validator.check(LanguagePattern.matches(language), "language", "validation.message.language.invalid")
    val text = body("text").flatMap(_.asString).getOrElse("")
    validator.check(text.nonEmpty, "text", "validation.message.text.required")
    validator.check(text.length <= 2000, "text", "validation.message.text.too-long")
    val description = body("description").flatMap(_.asString)
    validator.check(description.forall(_.length <= 1000), "description", "validation.message.description.too-long")
    validator.throwIfInvalid()
    MessageInput(key, language, text, description)
  }

  def validateReportInput(body: JsonObject): ReportInput = {
    val validator = Validator()
    val reason = body("reason").flatMap(_.asString)
    validator.check(
      reason.exists(Seq("spam", "offensive", "broken-link", "other").contains),
      "reason",
      "validation.report.reason.invalid"
    )
    val comment = body("comment").flatMap(_.asString)
    validator.check(comment.forall(_.length <= 1000), "comment", "validation.report.comment.too-long")
    validator.throwIfInvalid()
    ReportInput(reason.get, comment)
  }

  private def isHttpUrl(value: String): Boolean =
    Try(URI(value)).toOption.exists(uri =>
      (uri.getScheme == "http" || uri.getScheme == "https") && Option(uri.getHost).exists(_.nonEmpty)
    )
}
